include/swift/Runtime/KnownProtocolWitnessTables.h

```diff
@@ -0,0 +1,78 @@
+//
+// Copyright (c) 2019 Apple Inc. and the Swift project authors
+// Licensed under Apache License v2.0 with Runtime Library Exception
+//
+// See https://swift.org/LICENSE.txt for license information
+// See https://swift.org/CONTRIBUTORS.txt for the list of Swift project authors
+//
+//===----------------------------------------------------------------------===//
+//
+// Swift runtime support for known protocol witness tables and related items.
+//
+//===----------------------------------------------------------------------===//
+
+#include "swift/ABI/Metadata.h"
+
+namespace swift {
+
+// public protocol Equatable {}
+#define SWIFT_EQUATABLE_MANGLING SQ
+
+// == infix(T, T) -> Swift.Bool
+#define SWIFT_EQUAL_OPERATOR_MANGLING 2eeoiySbx_xtFZ
+
+#define PROTOCOL_DESCRIPTOR_MANGLING Mp
+#define BUILTIN_PROTOCOL_WITNESS_TABLE_MANGLING WB
+#define BUILTIN_PROTOCOL_WITNESS_MANGLING TB
+#define BUILTIN_PROTOCOL_CONFORMANCE_DESCRIPTOR_MANGLING Mb
+
+#define PROTOCOL_DESCRIPTOR_SYM(Proto) \
+          MANGLE_SYM(MANGLING_CONCAT2(Proto, PROTOCOL_DESCRIPTOR_MANGLING))
+
+// Note: This is currently only used for conformances found within Swift.
+//       (Hence the s module)
+#define BUILTIN_PROTOCOL_WITNESS_TABLE_SYM(Ty, Proto) \
+          MANGLE_SYM(MANGLING_CONCAT3(Ty, \
+                                      MANGLING_CONCAT2(Proto, s), \
+                                      BUILTIN_PROTOCOL_WITNESS_TABLE_MANGLING))
+
+// Note: This is currently only used for conformances found within Swift.
+//       (Hence the s module)
+#define BUILTIN_PROTOCOL_WITNESS_SYM(Ty, Proto, Func) \
+          MANGLE_SYM(MANGLING_CONCAT2(_PROTOCOL_CONFORMANCE_SYM(Ty, Proto, s), \
+                                      MANGLING_CONCAT3(Proto, Func, \
+                                        BUILTIN_PROTOCOL_WITNESS_MANGLING)))
+
+#define _PROTOCOL_CONFORMANCE_SYM(Ty, Proto, Mod) \
+          MANGLING_CONCAT3(Ty, Proto, Mod)
+
+// Note: This is currently only used for conformances found within Swift.
+//       (Hence the s module)
+#define BUILTIN_PROTOCOL_CONFORMANCE_DESCRIPTOR_SYM(Ty, Proto) \
+          MANGLE_SYM(MANGLING_CONCAT2(_PROTOCOL_CONFORMANCE_SYM(Ty, Proto, s), \
+                              BUILTIN_PROTOCOL_CONFORMANCE_DESCRIPTOR_MANGLING))
+
+struct _WitnessTable {
+  const ProtocolConformanceDescriptor *Conformance;
+  const void *Witness;
+};
+
+/// The protocol witness table for (): Swift.Equatable in Swift.
+SWIFT_RUNTIME_EXPORT
+const _WitnessTable BUILTIN_PROTOCOL_WITNESS_TABLE_SYM(EMPTY_TUPLE_MANGLING,
+                                                       SWIFT_EQUATABLE_MANGLING);
+
+/// The protocol witness for static Swift.Equatable.== infix(A, A) -> Swift.Bool
+/// in conformance (): Swift.Equatable in Swift.
+///
+/// Note: Empty tuples are not lowered as parameters, so it's fine to exclude.
+SWIFT_RUNTIME_EXPORT SWIFT_CC(swift)
+bool BUILTIN_PROTOCOL_WITNESS_SYM(EMPTY_TUPLE_MANGLING,
```

Comment:
I'm curious if this approach is a step towards enabling tuple conformances to arbitrary protocols? Or is it strictly limited to tuple conformances to specific known protocols?

Reply:
This approach essentially allows us to create conformances for any structural type to any protocol. We're only limiting this to certain protocols.

If you're referring to something like extending tuples and allowing them to declare conformances there, then there's more work needed there. I don't think tuple extensions are possible without parameterized extensions, but you'd also need variadic generics for those extensions to be really useful anyway. You know, the *usual*. x)

Reply:
Thanks for the response!

> This approach essentially allows us to create conformances for any structural type to any protocol. We're only limiting this to certain protocols.

I see. It seems to me that "*known* protocol witness tables" is inherently an approach for hardcoding conformances, so it's not user-extensible.

> If you're referring to something like extending tuples and allowing them to declare conformances there, then there's more work needed there. I don't think tuple extensions are possible without parameterized extensions, but you'd also need variadic generics for those extensions to be really useful anyway.

Yes, declaring tuple conformances in source code was my original question, sorry for not being more clear. Parameterized extensions make sense as an enabler:
```swift
extension<T: Equatable, U: Equatable> (T, U): Equatable { ... }
```

The ability to declare tuple conformances seems useful even without variadic generics: at least it becomes possible to express tuple conformances!

Reply:
I'm unsure whether this patch is a step towards that. I imagine once we have a real implementation of tuple extensions that we'd want to treat them as normal protocol conformances and emit actual witness table implementations. I did some reading on previous discussions of this and some were under the impression that these builtin conformances would be considered technical debt once we get a proper implementation, which I somewhat agree with. However, I'm also under the impression that doing this now is probably better than waiting for parameterized extensions (I'm ashamed I haven't worked on it in a while) and variadic generics. I can imagine that the work to get actual tuple conformances working after those land *shouldn't* be totally difficult.